src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+    if (!canUseGPSHeading
+#ifdef USE_MAG
+        && !compassIsHealthy()
+#endif
+        ) {
+        return false;
+    }
+    
+    if (!wasThrottleRaised()) {
+        return false;
+    }
+
+    if (isNewDataForPosHold()) {
+        posHold.gpsDataIntervalS = getGpsDataIntervalSeconds(); // interval for current GPS data value 0.01s to 1.0s
+        posHold.gpsDataFreqHz = 1.0f / posHold.gpsDataIntervalS;
+        if (posHold.sticksActive) {
+            // if a Position Hold deadband is set, and sticks are outside deadband, allow pilot control in angle mode
+            resetPositionControlParams(&posHold.NS);
+            resetPositionControlParams(&posHold.EW);
+            posHold.pidSumRoll = 0.0f;
+            posHold.pidSumPitch = 0.0f;
+        } else {
+            // first get xy distances from current location (gpsSol.llh) to target location
+            float nsDistance; // cm, steps of 11.1cm, North of target is positive
+            float ewDistance; // cm, steps of 11.1cm, East of target is positive
+            GPS_distances(&gpsSol.llh, &currentTargetLocation, &nsDistance, &ewDistance);
+            float distanceCm = sqrtf(sq(nsDistance) + sq(ewDistance));
+    
+            posHold.NS.distance = nsDistance;
+            posHold.EW.distance = ewDistance;
+    
+            posHold.pt1Gain = pt1FilterGain(posHold.lpfCutoff, posHold.gpsDataIntervalS);
+            const float leak = 1.0f - 0.4f * posHold.gpsDataIntervalS; // gpsDataIntervalS is not more than 1.0s
+    
+            // ** Sanity check **
+            // larger threshold if faster at start
+            if (posHold.NS.isStarting || posHold.EW.isStarting) {
+                posHold.sanityCheckDistance = gpsSol.groundSpeed > 1000 ? gpsSol.groundSpeed : 1000.0f;
+                // 1s of flight at current speed or 10m, in cm
+            }
+            // primarily to detect flyaway from no Mag or badly oriented Mag
+            // but must accept some overshoot at the start, especially if entering at high speed
+            if (distanceCm > posHold.sanityCheckDistance) {
+                return false; // must stay within 10m or probably flying away
+                // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+                // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+                // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+            }
+    
+            vectors_t *vectors[] = { &posHold.NS, &posHold.EW };
+            for (int i = 0; i < 2; i++) {
+                vectors_t *latLong = vectors[i];
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+    
+                // ** P **
+                float pidP = latLong->distance * positionPidCoeffs.Kp;
+    
+                // ** I **
+                if (!latLong->isStarting){
```

Comment:
Why is state for N different from E ? THis should be independent of rotation.

Usually such special-case handling works for expected case only, making all other cases worse. If you want to stop first, it would be better use cascaded controller - first set desired speed to zero and ignore position. After slowing down enough, start feeding position feedback into speed controller.

Reply:
It works best for each axis to reset whenever it's pidSum crosses zero, because then the crossover to the new target happens independently and smoothly.  Each of the vectors pidSums will cross zero at different times.

Originally I tracked the minimum GPS speed (combined vector speed) and stopped when it was lowest.  This wasn't very good, there was always an abrupt jump.

When I changed to consider two axes independently, I tried the same approach of a minimum abs velocity per-axis.  The problem is that 'minimum' velocity has it's own assumption - that there will be a minimum.  And we can't always be sure that velocity will ever be zero, or even cross zero. 

After a lot of testing, I realised that this method provides the smoothest transition.  That's because when the pidSum is zero, the requested attitude is zero, and we can transfer to the new target location, which gives us a zero new pidSum immediately on the transition.  Therefore no abrupt change in angle setpoint on the acceptance of the new location.  

Even if this fails, it doesn't really matter.  The quad will then keep trying to get to the original target location, that's all  At some point when the pidSum reverses, the target resets to whatever that point is.  Either way it is chasing a target.

If mag is wrong, or if the wind is so strong that the PIDs cannot win, and they keep growing, then the original target location (the target at initiation) will remain.  This is not so bad.

At this point the current approach works best, of the methods I tried so far, during the landing phase, where the pilot makes many small adjustments with the stick, and it is swapping target points in and out, and we need those swaps to be very smooth at this time.

Reply:
My main problem with this is that poshold behaves differently when flying north (one axis is latched almost immediately, possibly just because of noise) and north-east.

> After a lot of testing, I realised that this method provides the smoothest transition. That's because when the pidSum is zero, the requested attitude is zero, and we can transfer to the new target location, which gives us a zero new pidSum immediately on the transition. Therefore no abrupt change in angle setpoint on the acceptance of the new location.

The smooth transition is caused by filter reset. It at best a desirable glitch - both acceleration and velocity are usually nonzero when pidsum==0;

Poshold target is chosen some distance from point it was engaged (and it is different for N vs NE direction).

Did you try selecting target some distance in flight direction? Possibly assuming constant deceleration. 
Or rotate whole system so that one pid is forward (using GPS speed) and second is sideways (latched immediately)

Reply:
> Usually such special-case handling works for expected case only, making all other cases worse. If you want to stop first, it would be better use cascaded controller - first set desired speed to zero and ignore position. After slowing down enough, start feeding position feedback into speed controller.

This is how most autopilots do things as well. Only truly being in "Position mode" when the sticks are essentially centered and otherwise being in "Velocity mode". Although it can be beneficial to still feed some amount of position mode into the controller so that straight forward flight will tend to not drift side to side.